lib/stages/ReadGain.cpp

```diff
@@ -0,0 +1,240 @@
+    default_gains = config.get_default<std::vector<float>>(unique_name, "frb_missing_gains", dg);
+
+    metadata_buf = get_buffer("in_buf");
+    register_consumer(metadata_buf, unique_name.c_str());
+    metadata_buffer_id = 0;
+    metadata_buffer_precondition_id = 0;
+    freq_idx = -1;
+    freq_MHz = -1;
+
+    // Gain for FRB
+    gain_frb_buf = get_buffer("gain_frb_buf");
+    gain_frb_buf_id = 0;
+    register_producer(gain_frb_buf, unique_name.c_str());
+    update_gains_frb = true;
+
+    // Gain for PSR
+    gain_psr_buf = get_buffer("gain_psr_buf");
+    gain_psr_buf_id = 0;
+    register_producer(gain_psr_buf, unique_name.c_str());
+    update_gains_psr = true;
+
+    first_pass = true;
+
+    using namespace std::placeholders;
+
+    // listen for gain updates FRB
+    _gain_dir_frb = config.get_default<std::string>(unique_name, "updatable_config/gain_frb", "");
+    if (_gain_dir_frb.length() > 0)
+        configUpdater::instance().subscribe(
+            config.get<std::string>(unique_name, "updatable_config/gain_frb"),
+            std::bind(&ReadGain::update_gains_frb_callback, this, _1));
+
+    // listen for gain updates PSR
+    configUpdater::instance().subscribe(
+        config.get<std::string>(unique_name, "updatable_config/gain_psr"),
+        std::bind(&ReadGain::update_gains_psr_callback, this, _1));
+}
+
+bool ReadGain::update_gains_frb_callback(nlohmann::json& json) {
+    {
+        std::lock_guard<std::mutex> lock(mux);
+        update_gains_frb = true;
+    }
+    cond_var.notify_all();
+
+    try {
+        _gain_dir_frb = json.at("frb_gain_dir");
+    } catch (std::exception& e) {
+        WARN("[FRB] Fail to read gain_dir {:s}", e.what());
+        return false;
+    }
+    INFO("[ReadGain] updated gain with {:s}============update_gains={:d}", _gain_dir_frb,
+         update_gains_frb);
+    return true;
+}
+
+bool ReadGain::update_gains_psr_callback(nlohmann::json& json) {
+    {
+        std::lock_guard<std::mutex> lock(mux);
+        update_gains_psr = true;
+    }
+    cond_var.notify_all();
+
+    try {
+        _gain_dir_psr = json.at("pulsar_gain_dir").get<std::vector<string>>();
+        INFO("[PSR] Updating gains from {:s} {:s} {:s} {:s} {:s} {:s} {:s} {:s} {:s} {:s}",
+             _gain_dir_psr[0].c_str(), _gain_dir_psr[1].c_str(), _gain_dir_psr[2].c_str(),
+             _gain_dir_psr[3].c_str(), _gain_dir_psr[4].c_str(), _gain_dir_psr[5].c_str(),
+             _gain_dir_psr[6].c_str(), _gain_dir_psr[7].c_str(), _gain_dir_psr[8].c_str(),
+             _gain_dir_psr[9].c_str());
+    } catch (std::exception const& e) {
+        WARN("[PSR] Fail to read gain_dir {:s}", e.what());
+        return false;
+    }
+    return true;
+}
+
+void ReadGain::main_thread() {
+
+    if (first_pass) {
+        first_pass = false;
+        uint8_t* frame =
+            wait_for_full_frame(metadata_buf, unique_name.c_str(), metadata_buffer_precondition_id);
+        if (frame == NULL)
+            goto end_loop;
```

Comment:
Avoid `goto` unless it's needed.
```suggestion
            return;
```